File: src/content/docs/topics/cs/leetcode-150/arrays-and-hashing/238-product-of-array-except-self.py
def product_except_self(nums: list[int]) -> list[int]:
    pass  # TODO: implement

def _run_tests():
    assert product_except_self([1, 2, 3, 4]) == [24, 12, 8, 6]
    assert product_except_self([-1, 1, 0, -3, 3]) == [0, 0, 9, 0, 0]
    assert product_except_self([1, 1]) == [1, 1]
    assert product_except_self([2, 3]) == [3, 2]
    assert product_except_self([1, 0]) == [0, 1]
    # --- large-input timing ---
    import time as _t
    _t0 = _t.perf_counter()
    product_except_self(list(range(1, 10001)))
    _ms = (_t.perf_counter() - _t0) * 1000
    print(f'perf product_except_self n=10000: {_ms:.1f}ms')
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/238-product-of-array-except-self.swift
// LEETCODE_TYPE: Solution
func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ message: String = "",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard actual == expected else {
        let detail = message.isEmpty ? "values differ" : message
        fatalError("\(file):\(line): \(detail). Expected \(expected), got \(actual)")
    }
}

func expectTrue(
    _ condition: @autoclosure () -> Bool,
    _ message: String = "expected true",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard condition() else {
        fatalError("\(file):\(line): \(message)")
    }
}

func reportSuccess() {
    print("All Swift tests passed")
}

final class Solution {
    func productExceptSelf(_ nums: [Int]) -> [Int] {
        // TODO: Implement
        fatalError("TODO: Implement")
    }
}

func runTests() {
    // TEST_VECTORS_BEGIN sha256:c18be99084dbf1d96fb5ca0a038674f10088ca91fcc46d9a3eb94c12c217506a
    expectEqual(Solution().productExceptSelf([1, 2, 3, 4]), [24, 12, 8, 6], "canonical")
    expectEqual(Solution().productExceptSelf([-1, 1, 0, -3, 3]), [0, 0, 9, 0, 0], "contains-zero")
    expectEqual(Solution().productExceptSelf([2, 3]), [3, 2], "two")
    expectEqual(Solution().productExceptSelf([1, 0]), [0, 1], "one-zero")
    expectEqual(Solution().productExceptSelf([0, 0, 2]), [0, 0, 0], "two-zeros")
    // EXCLUDED_VECTOR single: [[5]] | The input contains at least two values.
    // TEST_VECTORS_END
    reportSuccess()
}

runTests()
